src/client/builtin_scene/css_border_data_texture.cpp

```diff
@@ -0,0 +1,191 @@
+      // Columns 1-4: border colors for each side
+      for (int side = 0; side < 4; ++side)
+      {
+        size_t colorOffset = rowOffset + (side + 1) * 4;
+        textureData_[colorOffset + 0] = borderColors[side].r;
+        textureData_[colorOffset + 1] = borderColors[side].g;
+        textureData_[colorOffset + 2] = borderColors[side].b;
+        textureData_[colorOffset + 3] = borderColors[side].a;
+      }
+    }
+
+    // Upload data to GPU
+    glContext_->activeTexture(WebGLTextureUnit::kTexture1);
+    glContext_->bindTexture(WebGLTextureBindingTarget::k2D, borderDataTexture_);
+    glContext_->texSubImage2D(WebGLTextureBindingTarget::k2D,
+                              0, // mip level
+                              0, 0, // x, y offset
+                              5, // width
+                              instanceCount, // height
+                              WebGLTextureFormat::kRGBA, // format
+                              WebGLPixelType::kFloat, // type
+                              textureData_.data()); // data
+
+    isDirty_ = false;
+  }
+
+  void CSSBorderDataTexture::bind(WebGLTextureUnit textureUnit)
+  {
+    if (TR_UNLIKELY(!isInitialized()))
+      return;
+
+    glContext_->activeTexture(textureUnit);
+    glContext_->bindTexture(WebGLTextureBindingTarget::k2D, borderDataTexture_);
+  }
+
+  void CSSBorderDataTexture::ensureTextureSize(size_t instanceCount)
+  {
+    if (instanceCount <= currentTextureHeight_)
+      return;
+
+    // Resize texture data buffer
+    size_t newHeight = instanceCount;
+    textureData_.resize(5 * 4 * newHeight, 0.0f);
+
+    // Reallocate GPU texture
+    glContext_->activeTexture(WebGLTextureUnit::kTexture1);
+    glContext_->bindTexture(WebGLTextureBindingTarget::k2D, borderDataTexture_);
+    glContext_->texImage2D(WebGLTextureBindingTarget::k2D,
+                           0, // mip level
+                           WEBGL2_RGBA32F, // internal format
+                           5, // width
+                           newHeight, // height
+                           0, // border
+                           WebGLTextureFormat::kRGBA, // format
+                           WebGLPixelType::kFloat, // type
+                           nullptr); // data (will be uploaded via texSubImage2D)
+
+    currentTextureHeight_ = newHeight;
+    isDirty_ = true;
+  }
+
+  void CSSBorderDataTexture::extractInstanceBorderData(const Instance& instance,
+                                                        glm::vec4& borderWidth,
+                                                        glm::vec4 borderColors[4])
+  {
+    // Extract border width from instance
+    // Currently the Instance class has borderWidths_ and borderColors_[4] fields
+    // but the setter methods are commented out. For now, we'll use default values
+    // and set up the infrastructure for when border data is properly populated.
+    
+    // TODO: Once Instance border data setters are implemented, extract actual data
```

Comment:
The `Instance` class has these border fields, you can help implementing getters for this.